pytorch_lightning/core/lightning.py

```diff
@@ -1799,42 +1713,21 @@ def to_torchscript(
 
         return torchscript_module
 
-    @property
-    def hparams(self) -> Union[AttributeDict, dict, Namespace]:
-        if not hasattr(self, "_hparams"):
-            self._hparams = AttributeDict()
-        return self._hparams
+    def add_datamodule_hparams(self, datamodule: LightningDataModule):
+        """Add the hparams of a LightningDataModule to the hparams and hparams_initial of this module."""
+        if not hasattr(datamodule, 'hparams'):
+            return
 
-    @property
-    def hparams_initial(self) -> AttributeDict:
-        if not hasattr(self, "_hparams_initial"):
-            return AttributeDict()
-        # prevent any change
-        return copy.deepcopy(self._hparams_initial)
-
-    @hparams.setter
-    def hparams(self, hp: Union[dict, Namespace, Any]):
-        hparams_assignment_name = self.__get_hparams_assignment_variable()
-        self._hparams_name = hparams_assignment_name
-        self._set_hparams(hp)
-        # this resolves case when user does not uses `save_hyperparameters` and do hard assignement in init
-        if not hasattr(self, "_hparams_initial"):
-            self._hparams_initial = copy.deepcopy(self._hparams)
-
-    def __get_hparams_assignment_variable(self):
-        """"""
-        """
-        looks at the code of the class to figure out what the user named self.hparams
-        this only happens when the user explicitly sets self.hparams
-        """
-        try:
-            class_code = inspect.getsource(self.__class__)
-            lines = class_code.split("\n")
-            for line in lines:
-                line = re.sub(r"\s+", "", line, flags=re.UNICODE)
-                if ".hparams=" in line:
-                    return line.split("=")[1]
-        except Exception as e:
-            return "hparams"
-
-        return None
+        hparams = self._to_hparams_dict(datamodule.hparams)
+        if not hasattr(self, '_hparams'):
+            self._hparams = hparams
+            self._hparams_initial = hparams
+        else:
+            colliding_keys = [key for key in hparams.keys() if key in self.hparams]
+            colliding_keys += [key for key in hparams.keys() if key in self.hparams_initial]
+            colliding_keys = set(colliding_keys)
+            if colliding_keys:
+                raise ValueError(f'Error while adding datamodule hparams: '
+                                 f'the keys {colliding_keys} are already present in the model hparams.')
```

Comment:
```suggestion
                raise ValueError(
                    f'Error while adding datamodule hparams: the keys {colliding_keys} are already present in the model hparams.'
                )
```

Reply:
Does this mean I can't have something like the same `input_dim` key in both model (for setting first layer dims) and datamodule (for setting resize dims) hparams?

Reply:
Yeah, that's the idea. Previously I added a prefix to the data hyperparameters but it was getting convoluted with all the duplicate hparams. I could change it back again?